kube-runtime/src/controller/mod.rs

```diff
@@ -755,6 +848,98 @@ where
+    /// # use std::sync::Arc;
+    /// # type CustomResource = ConfigMap;
+    /// # async fn reconcile(_: Arc<CustomResource>, _: Arc<()>) -> Result<Action, Error> { Ok(Action::await_change()) }
+    /// # fn error_policy(_: Arc<CustomResource>, _: &kube::Error, _: Arc<()>) -> Action { Action::await_change() }
+    /// fn mapper(_: DaemonSet) -> Option<ObjectRef<CustomResource>> { todo!() }
+    /// # async fn doc(client: kube::Client) {
+    /// let api: Api<DaemonSet> = Api::all(client.clone());
+    /// let cr: Api<CustomResource> = Api::all(client.clone());
+    /// let daemons = watcher(api, watcher::Config::default())
+    ///     .touched_objects()
+    ///     .predicate_filter(predicates::generation);
+    ///
+    /// Controller::new(cr, watcher::Config::default())
+    ///     .watches_stream(daemons, mapper)
+    ///     .run(reconcile, error_policy, Arc::new(()))
+    ///     .for_each(|_| std::future::ready(()))
+    ///     .await;
+    /// # }
+    /// ```
+    #[cfg(feature = "unstable-runtime-stream-control")]
+    #[must_use]
+    pub fn watches_stream<Other, I>(
+        self,
+        trigger: impl Stream<Item = Result<Other, watcher::Error>> + Send + 'static,
+        mapper: impl Fn(Other) -> I + Sync + Send + 'static,
+    ) -> Self
+    where
+        Other: Clone + Resource + DeserializeOwned + Debug + Send + 'static,
+        Other::DynamicType: Default + Debug + Clone,
+        I: 'static + IntoIterator<Item = ObjectRef<K>>,
+        I::IntoIter: Send,
+    {
+        self.watches_stream_with(trigger, mapper, Default::default())
+    }
+
+    /// Trigger the reconciliation process for a stream of `Other` objects related to a `K`
+    ///
+    /// Same as [`Controller::owns`], but instead of an `Api`, a stream of resources is used.
+    /// This allows for customized and pre-filtered watch streams to be used as a trigger,
+    /// as well as sharing input streams between multiple controllers.
+    ///
+    /// **NB**: This is constructor requires an [`unstable`](https://github.com/kube-rs/kube/blob/main/kube-runtime/Cargo.toml#L17-L21) feature.
+    ///
+    /// Same as [`Controller::watches_stream`], but accepts a `DynamicType` so it can be used with dynamic resources.
+    #[cfg(feature = "unstable-runtime-stream-control")]
+    #[must_use]
+    pub fn watches_stream_with<Other, I>(
+        mut self,
+        trigger: impl Stream<Item = Result<Other, watcher::Error>> + Send + 'static,
+        mapper: impl Fn(Other) -> I + Sync + Send + 'static,
+        dyntype: Other::DynamicType,
+    ) -> Self
+    where
+        Other: Clone + Resource + DeserializeOwned + Debug + Send + 'static,
+        Other::DynamicType: Debug + Clone,
+        I: 'static + IntoIterator<Item = ObjectRef<K>>,
+        I::IntoIter: Send,
+    {
+        let other_watcher = trigger_with(trigger, move |obj| {
+            let watched_obj_ref = ObjectRef::from_obj_with(&obj, dyntype.clone()).erase();
+            mapper(obj)
```

Comment:
This is very similar to `trigger_owners`, can't we implement `trigger_owners` using this by just supplying a mapper?

Maybe move this out into a standalone `trigger_*` function and implement `trigger_owners` using it?

Reply:
Yeah, you're right. The pattern is that most of the bodies of the `Controller::methods` are factored out into trigger helpers in:
https://github.com/kube-rs/kube/blob/36dcdaaad178d4f6cbe0bd7513d9608bfcf09bb6/kube-runtime/src/controller/mod.rs#L87-L150

ended up doing https://github.com/kube-rs/kube/pull/1187/commits/fc77780f8bbcebb32633f8a1e2c8fe4de4749f3a which has a fairly horrendous generic signature, but it's nice to have it out of the imperative `Controller` logic.

I think long term all these trigger helpers can be turned into `WatchStreamExt` methods as that will make it chain nicer. Will open an issue for this.